File: client/models/Spirit.ts
export interface SpiritRawData {
    id: string | null;
    name: string | null;
    description: string | null;
    primaryType: string | null;
    secondaryType: string | null;
    originalImageDownloadUrl: string | null;
    generatedImageDownloadUrl: string | null;

    agility: number | null;
    arcana: number | null;
    aura: number | null;
    charisma: number | null;
    endurance: number | null;
    height: number | null;
    weight: number | null;
    intimidation: number | null;
    luck: number | null;
    hitPoints: number | null;
    strength: number | null;
    toughness: number | null;
  }
  
export interface Spirit {
    id: string;
    name: string;
    description: string;
    primaryType: string;
    secondaryType: string;
    originalImageDownloadUrl: string;
    generatedImageDownloadUrl: string;

    agility: number;
    arcana: number;
    aura: number;
    charisma: number;
    endurance: number;
    height: number;
    weight: number;
    intimidation: number;
    luck: number;
    maxHitPoints: number;
    currentHitPoints: number;
    strength: number;
    toughness: number;
  }

// Utility function to check if a SpiritRawData is complete.
export function isCompleteSpirit(spirit: SpiritRawData): boolean {
  return spirit.id != null &&
    spirit.name != null &&
    spirit.description != null &&
    spirit.primaryType != null &&
    spirit.secondaryType != null &&
    spirit.originalImageDownloadUrl != null &&
    spirit.generatedImageDownloadUrl != null &&
    spirit.agility != null &&
    spirit.arcana != null &&
    spirit.aura != null &&
    spirit.charisma != null &&
    spirit.endurance != null &&
    spirit.height != null &&
    spirit.weight != null &&
    spirit.intimidation != null &&
    spirit.luck != null &&
    spirit.hitPoints != null &&
    spirit.strength != null &&
    spirit.toughness != null
}

export function convertSpiritRawDataToSpirit(spirit: SpiritRawData): Spirit {
  if (!isCompleteSpirit(spirit)) {
    throw new Error('Spirit data is incomplete');
  }
  return {
    id: spirit.id as string,
    name: spirit.name as string,
    description: spirit.description as string,
    primaryType: spirit.primaryType as string,
    secondaryType: spirit.secondaryType as string,
    originalImageDownloadUrl: spirit.originalImageDownloadUrl as string,
    generatedImageDownloadUrl: spirit.generatedImageDownloadUrl as string,
    agility: spirit.agility as number,
    arcana: spirit.arcana as number,
    aura: spirit.aura as number,
    charisma: spirit.charisma as number,
    endurance: spirit.endurance as number,
    height: spirit.height as number,
    weight: spirit.weight as number,
    intimidation: spirit.intimidation as number,
    luck: spirit.luck as number,
    maxHitPoints: spirit.hitPoints as number,
    currentHitPoints: spirit.hitPoints as number,
    strength: spirit.strength as number,
    toughness: spirit.toughness as number  }
}

File: server/models/spirit.go
package models

import (
	"context"
)

// This is the model of the Spirit object that will be returned to the client.
type Spirit struct {
	ID                *string `json:"id"`
	Name              *string `json:"name"`
	Description       *string `json:"description"`
	PrimaryType       *string `json:"primaryType"`
	SecondaryType     *string `json:"secondaryType"`
	OriginalImageURL  *string `json:"originalImageDownloadUrl"`
	GeneratedImageURL *string `json:"generatedImageDownloadUrl"`
	Moves             []*Move `json:"moves"`

	Agility      *int `json:"agility"`
	Arcana       *int `json:"arcana"`
	Aura         *int `json:"aura"`
	Charisma     *int `json:"charisma"`
	Endurance    *int `json:"endurance"`
	Height       *int `json:"height"`
	Weight       *int `json:"weight"`
	Intimidation *int `json:"intimidation"`
	Luck         *int `json:"luck"`
	Strength     *int `json:"strength"`
	Toughness    *int `json:"toughness"`
	HitPoints    *int `json:"hitPoints"`
}

type StorageInterface interface {
	GetDownloadURL(ctx context.Context, bucketName, objectName string) (string, error)
}

type DatastoreInterface interface {
	GetDocumentsByIds(ctx context.Context, collectionName string, ids []string) ([]map[string]interface{}, error)
}

func getMoves(ctx context.Context, datastoreClient DatastoreInterface, moveIds []string) []*Move {
	moveDocs, _ := datastoreClient.GetDocumentsByIds(ctx, "moves", moveIds)
	if moveDocs == nil {
		return nil
	}

	var moves []*Move
	for _, doc := range moveDocs {
		moves = append(moves, BuildMovefromDocData(doc))
	}

	return moves
}

func BuildSpiritfromDocData(ctx context.Context, storageClient StorageInterface, doc map[string]interface{}, datastoreClient DatastoreInterface) Spirit {
	id := GetOptionalStringField(doc, "id")
	name := GetOptionalStringField(doc, "name")
	description := GetOptionalStringField(doc, "description")
	primaryType := GetOptionalStringField(doc, "primaryType")
	secondaryType := GetOptionalStringField(doc, "secondaryType")

	var originalUrl, generatedUrl *string
	originalUrl = getImageURL(ctx, storageClient, doc, "originalImageFilePath")
	generatedUrl = getImageURL(ctx, storageClient, doc, "generatedImageFilePath")

	moveIds := GetOptionalStringArrayField(doc, "moveIds")
	moves := []*Move{}
	if moveIds != nil {
		moves = getMoves(ctx, datastoreClient, moveIds)
	} else {
		moves = nil
	}

	// Extract numeric fields and set default values if not present
	agility := GetOptionalIntField(doc, "agility")
	arcana := GetOptionalIntField(doc, "arcana")
	aura := GetOptionalIntField(doc, "aura")
	charisma := GetOptionalIntField(doc, "charisma")
	endurance := GetOptionalIntField(doc, "endurance")
	height := GetOptionalIntField(doc, "height")
	weight := GetOptionalIntField(doc, "weight")
	intimidation := GetOptionalIntField(doc, "intimidation")
	luck := GetOptionalIntField(doc, "luck")
	strength := GetOptionalIntField(doc, "strength")
	toughness := GetOptionalIntField(doc, "toughness")
	hitPoints := GetOptionalIntField(doc, "hitPoints")

	return Spirit{
		ID:                id,
		Name:              name,
		Description:       description,
		PrimaryType:       primaryType,
		SecondaryType:     secondaryType,
		OriginalImageURL:  originalUrl,
		GeneratedImageURL: generatedUrl,
		Moves:             moves,

		Agility:      agility,
		Arcana:       arcana,
		Aura:         aura,
		Charisma:     charisma,
		Endurance:    endurance,
		Height:       height,
		Weight:       weight,
		Intimidation: intimidation,
		Luck:         luck,
		Strength:     strength,
		Toughness:    toughness,
		HitPoints:    hitPoints,
	}
}
